internal/wal/main.go

```diff
@@ -0,0 +1,210 @@
+package wal
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/dicedb/dice/internal/cmd"
+	"google.golang.org/protobuf/proto"
+)
+
+type WAL struct {
+	file   *os.File
+	mutex  sync.Mutex
+	logDir string
+	ticker *time.Ticker
+	stopCh chan struct{}
+}
+
+func NewWAL(logDir string) (*WAL, error) {
+	wal := &WAL{
+		logDir: logDir,
+		ticker: time.NewTicker(1 * time.Minute),
+		stopCh: make(chan struct{}),
+	}
+
+	// Ensure the log directory exists
+	if err := os.MkdirAll(logDir, 0755); err != nil {
+		return nil, fmt.Errorf("failed to create log directory: %v", err)
+	}
+
+	if err := wal.rotateLogFile(); err != nil {
+		return nil, fmt.Errorf("failed to create initial log file: %v", err)
+	}
+
+	go wal.rotateLogPeriodically()
+	return wal, nil
+}
+
+// rotateLogFile closes the current WAL file and opens a new one with a timestamped name.
+func (w *WAL) rotateLogFile() error {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	if w.file != nil {
+		w.file.Close()
+	}
+
+	// Create new file with minute-level timestamp suffix
+	timestamp := time.Now().Format("20060102_1504")
+	filePath := filepath.Join(w.logDir, fmt.Sprintf("wal_%s.log", timestamp))
+	newFile, err := os.OpenFile(filePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return fmt.Errorf("failed to open new WAL file: %v", err)
+	}
+
+	w.file = newFile
+	return nil
+}
+
+func (w *WAL) rotateLogPeriodically() {
+	for {
+		select {
+		case <-w.ticker.C:
+			if err := w.rotateLogFile(); err != nil {
+				fmt.Printf("Error rotating log file: %v\n", err)
+			}
+		case <-w.stopCh:
+			return
+		}
+	}
+}
+
+// LogCommand serializes a WALLogEntry and writes it to the current WAL file.
+func (w *WAL) LogCommand(c *cmd.DiceDBCmd) {
+	w.mutex.Lock()
```

Comment:
LogCommand seems to be a high contention and compute intensive method that involves locks and marshalling(although we are using PROTOBUF). Having this as part of critical path may effect latency. Before adding this code I would suggest we run the memtier benchmarks.